--- fontconfig-2.2.94/src/fcfreetype.c.orig	2004-04-14 16:57:48.000000000 +0200
+++ fontconfig-2.2.94/src/fcfreetype.c	2004-04-25 11:35:22.032861928 +0200
@@ -1829,7 +1829,7 @@
     return FcFalse;
 }
 
-#define APPROXIMATELY_EQUAL(x,y) (ABS ((x) - (y)) <= MAX (ABS (x), ABS (y)) / 33)
+#define APPROXIMATELY_EQUAL(x,y) (FT_ABS ((x) - (y)) <= FT_MAX (FT_ABS (x), FT_ABS (y)) / 33)
 
 FcCharSet *
 FcFreeTypeCharSetAndSpacing (FT_Face face, FcBlanks *blanks, int *spacing)
@@ -2046,7 +2046,7 @@
 #endif
     if (fixed_advance)
 	*spacing = FC_MONO;
-    else if (dual_advance && APPROXIMATELY_EQUAL (2 * MIN (advance_one, advance_two), MAX (advance_one, advance_two)))
+    else if (dual_advance && APPROXIMATELY_EQUAL (2 * FT_MIN (advance_one, advance_two), FT_MAX (advance_one, advance_two)))
         *spacing = FC_DUAL;
     else
 	*spacing = FC_PROPORTIONAL;
