diff -urN aaa/fcpackage.2_0/fontconfig/setfontdirs fcpackage.2_0/fontconfig/setfontdirs
--- aaa/fcpackage.2_0/fontconfig/setfontdirs	Thu Aug  1 18:17:33 2002
+++ fcpackage.2_0/fontconfig/setfontdirs	Tue Oct 15 21:29:07 2002
@@ -11,7 +11,7 @@
 chmod +w fonts.conf
 ed fonts.conf << EOF
 /FONTPATH_END/a
-<!-- Font directory list configured on `date` -->
+<!-- Font directory list configured on `(LC_ALL=C date)` -->
 .
 +r $FONTDIRS
 a
