--- fontconfig-2.2.92/src/fcfreetype.c.orig	2003-09-06 21:40:41.000000000 +0200
+++ fontconfig-2.2.92/src/fcfreetype.c	2003-12-22 22:03:40.219493768 +0100
@@ -48,16 +48,16 @@
 #include <stdio.h>
 #include <string.h>
 #include "fcint.h"
-#include <freetype/freetype.h>
-#include <freetype/internal/ftobjs.h>
-#include <freetype/tttables.h>
-#include <freetype/ftsnames.h>
-#include <freetype/ttnameid.h>
-#include <freetype/t1tables.h>
+#include <ft2build.h>
+#include FT_FREETYPE_H
+#include FT_INTERNAL_OBJECTS_H
+#include FT_TRUETYPE_TABLES_H
+#include FT_SFNT_NAMES_H
+#include FT_TRUETYPE_IDS_H
+#include FT_TYPE1_TABLES_H
 
 #if HAVE_FT_GET_BDF_PROPERTY
-#include <freetype/ftbdf.h>
-#include <freetype/ftmodule.h>
+#include FT_BDF_H
 #define HAS_BDF_PROPERTY(f) ((f) && (f)->driver && \
 			     (f)->driver->root.clazz->get_interface)
 #define MY_Get_BDF_Property(f,n,p) (HAS_BDF_PROPERTY(f) ? \
--- fontconfig-2.2.92/fontconfig/fcfreetype.h.orig	2003-07-20 19:14:47.000000000 +0200
+++ fontconfig-2.2.92/fontconfig/fcfreetype.h	2003-12-22 21:59:51.199310128 +0100
@@ -24,7 +24,8 @@
 
 #ifndef _FCFREETYPE_H_
 #define _FCFREETYPE_H_
-#include <freetype/freetype.h>
+#include <ft2build.h>
+#include FT_FREETYPE_H
 
 _FCFUNCPROTOBEGIN
 
